File: backend/src/main/java/com/skewwhiffy/auraltester/notation/model/note/Accidental.java
package com.skewwhiffy.auraltester.notation.model.note;

import org.apache.logging.log4j.util.Strings;

public record Accidental(int offset) {
    public String getAbc() {
        if (offset == 0) {
            return "";
        }
        if (offset < 0) {
            return Strings.repeat("_", -offset);
        }
        return Strings.repeat("^", offset);
    }

    /*
  val displayString: String = when {
    offset < 0 -> "b".repeat(-offset)
    offset % 2 == 0 -> "x".repeat(offset / 2)
    else -> "x".repeat((offset - 1) / 2) + "#"
  }
  */

    public Accidental flatten() {
        return new Accidental(offset - 1);
    }

    public Accidental sharpen() {
        return new Accidental(offset + 1);
    }

    public static Accidental getNatural() {
        return new Accidental(0);
    }

    public static Accidental getFlat() {
        return new Accidental(-1);
    }

    public static Accidental getSharp() {
        return new Accidental(1);
    }
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/note/Accidental.kt
package com.skewwhiffy.auraltester.notation.model.note

import com.skewwhiffy.auraltester.dao.AccidentalDao

data class Accidental(val offset: Int) {
    companion object {
        val natural: Accidental
            get() {
                return Accidental(0)
            }

        val flat: Accidental
            get() {
                return Accidental(-1)
            }

        val sharp: Accidental
            get() {
                return Accidental(1)
            }
    }

}

val Accidental.dao
    get() = AccidentalDao(offset)

val Accidental.abc
    get() = if (offset == 0) ""
    else if (offset < 0) "_".repeat(-offset)
    else "^".repeat(offset)

val Accidental.displayString: String
    get() {
        if (offset < 0) {
            return "b".repeat(-offset)
        }
        if (offset % 2 == 0) {
            return "x".repeat(offset / 2)
        }
        return "x".repeat((offset - 1) / 2) + "#"
    }

val Accidental.flatten
    get() = Accidental(offset - 1)

val Accidental.sharpen
    get() = Accidental(offset + 1)
